tests/test_http_request.py

```diff
@@ -1426,6 +1426,62 @@ def test_from_response_valid_form_methods(self):
             r = self.request_class.from_response(response)
             self.assertEqual(r.method, expected)
 
+    def test_form_response_with_invalid_formdata_type_error(self):
+        """Test that a ValueError is raised for non-iterable and non-dict formdata input"""
+        response = _buildresponse(
+            """<html><body>
+            <form action="/submit" method="post">
+                <input type="text" name="test" value="value">
+            </form>
+            </body></html>"""
+        )
+        with self.assertRaises(ValueError) as context:
+            FormRequest.from_response(response, formdata=123)
+
+        self.assertIn(
+            "formdata should be a dict or iterable of tuples", str(context.exception)
+        )
+
+    def test_form_response_with_custom_invalid_formdata_value_error(self):
+        """Test that a ValueError is raised for fault-inducing iterable formdata input"""
+        response = _buildresponse(
+            """<html><body>
+                <form action="/submit" method="post">
+                    <input type="text" name="test" value="value">
+                </form>
+            </body></html>"""
+        )
+
+        class CustomFormdata:
+            def __iter__(self):
+                raise ValueError("Custom iteration error for testing")
```

Comment:
This feels too artificial, I think. `("a",)` should work the same as `CustomFormdata()` below.